### PR TITLE
mpir/pmi: Add PMIx implementation of publish/lookup functionality

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Add implementation for PMIx publish/lookup. Make a minor change to the test infrastructure in order to exercise the new code.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

File: src/nameserv/pmi/pmi_nameserv.c [hunk 1]
```diff
@@ -109,162 +61,3 @@ int MPID_NS_Free(MPID_NS_Handle * handle_ptr)
     MPL_UNREFERENCED_ARG(handle_ptr);
     return 0;
 }
-
-#if 0
```
Comment (raffenet) [inline src/nameserv/pmi/pmi_nameserv.c:113]: just realized this removal got lumped into this commit. will split out once tests finish.

File: test/mpi/configure.ac [hunk 2]
```diff
@@ -512,21 +512,6 @@ AC_SUBST(comm_overlap)
...
 threadsdir="threads"
-if test "$enable_threads" != "no" ; then
-    AC_MSG_CHECKING([whether MPI_THREAD_MULTIPLE is supported])
-    AC_RUN_IFELSE([
-            AC_LANG_PROGRAM([#include "mpi.h"],[
-                int provided;
-                MPI_Init_thread(0,0, MPI_THREAD_MULTIPLE, &provided);
-                MPI_Finalize();
-                if (provided != MPI_THREAD_MULTIPLE) {
-                    exit(1);
```
Comment (hzhou) [inline test/mpi/configure.ac:523]: I think `return 1` will have the same effect.

Comment (raffenet) [reply test/mpi/configure.ac:523]: It does, but there are further complications with the publish test when run as a singleton:
pmi-1
```
Abort(402268449) on node 0 (rank 0 in comm 0): Fatal error in internal_Publish_name: Invalid service name (see MPI_Publish_name), error stack:
internal_Publish_name(72253): MPI_Publish_name(service_name=servname, MPI_INFO_NULL, port_name=otherhost:122) failed
MPIR_pmi_publish(1081)......: Lookup failed for service name servname
```
pmi-2
```
[unset]: PMIU_write error; fd=-1 buf=:    14cmd=job-getid;:
system msg for write_line failure : Bad file descriptor
[2]    85505 segmentation fault  ./a.out
```
pmix (singleton is broken, i have another patch to fix)
```
[2]    28193 segmentation fault  ./a.out
```

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

**Review by hzhou — APPROVED** [review]
LGTM

**Pull request merged**